qp_shotgun/humann2/humann2.py

```diff
@@ -11,23 +11,225 @@
+
+    # check that there aren't both paired and single sequences
+    if len(single) > 0 and len(fwd_paired) > 0:
+        raise ValueError('There are both paired-end and single-end sequences.')
+
+    # fill out unused seq file types with None and check that there exist files
+    if len(fwd_paired) > 0:
+        single = [None] * len(fwd_paired)
+    elif len(single) > 0:
+        fwd_paired = [None] * len(single)
+        fwd_unpaired = [None] * len(single)
+        rev_paired = [None] * len(single)
+        rev_unpaired = [None] * len(single)
+    else:
+        raise ValueError('There are no *.fastq.gz files in the artifact')
+
+    # make the 5-tuple of sorted sequence filepaths
+    fwd_paired.sort()
+    rev_paired.sort()
+    fwd_unpaired.sort()
+    rev_unpaired.sort()
+    single.sort()
+
+    seq_files = zip(fwd_paired, rev_paired, fwd_unpaired, rev_unpaired, single)
+
+    # get run prefixes
+    # These are prefixes that should match uniquely to forward reads
+    # sn_by_rp is dict of samples keyed by run prefixes
+    sn_by_rp = get_sample_names_by_run_prefix(map_file)
+
+    # make sets
+    read_sets = []
+    used_prefixes = set()
+
+    raised_ValueError = ''
+
+    for f_p, r_p, f_u, r_u, s in seq_files:
+        # pick file basename
+        if f_p is None:
+            fn = basename(s)
+        else:
+            fn = basename(f_p)
+
+        # iterate over run prefixes and make sure only one matches
+        run_prefix = None
+        for rp in sn_by_rp:
+            if fn.startswith(rp) and run_prefix is None:
+                run_prefix = rp
+            elif fn.startswith(rp) and run_prefix is not None:
+                raised_ValueError.append('Multiple run prefixes match this '
+                                         'file: %s\n\n' % fn)
+
+        # make sure that we got one matching run prefix:
+        if run_prefix is None:
+            raised_ValueError.append('No run prefix matching this read file: '
+                                     '%s\n\n' % fn)
+
+        if run_prefix in used_prefixes:
+            raised_ValueError.append('This run prefix matches multiple read '
+                                     ' files: %s\n\n' % run_prefix)
+
+        # if paired, check that all files match run prefix
+        if s is None:
+            if not (basename(r_p).startswith(run_prefix) and
+                    basename(f_u).startswith(run_prefix) and
+                    basename(r_u).startswith(run_prefix)):
+                raised_ValueError.append('Not all read files match run prefix.'
+                                         '\nRun prefix: %s\nForward paired: '
+                                         '%s\nReverse paired: %s\nForward '
+                                         'unpaired: %s\nReverse unpaired: %s\n'
+                                         % (run_prefix, f_p, r_p, f_u, r_u))
+
+        if raised_ValueError:
+            raise ValueError(raised_ValueError)
```

Comment:
Changed back to per-check raises. It turns out that some errors will lead to run_prefix remaining None, which causes a TypeError. Seems better to me to just raise the original causal error.